File: Server/src/main/scala/jk_5/nailed/server/command/CommandGamemode.java
package jk_5.nailed.server.command;

import java.util.*;

import net.minecraft.command.*;
import net.minecraft.entity.player.*;

import jk_5.nailed.api.*;
import jk_5.nailed.api.map.Map;
import jk_5.nailed.api.player.*;

/**
 * No description given
 *
 * @author jk-5
 */
public class CommandGamemode extends NailedCommand {

    public CommandGamemode() {
        super("gamemode");
    }

    @Override
    public List<String> getAliases() {
        return Arrays.asList("gm");
    }

    @Override
    public String getCommandUsage(ICommandSender icommandsender) {
        return "commands.gamemode.usage";
    }

    @Override
    public void processCommandWithMap(ICommandSender sender, Map map, String[] args) {
        if(args.length == 0){
            // /gm
            if(sender instanceof EntityPlayer){
                Player target = NailedAPI.getPlayerRegistry().getPlayer((EntityPlayer) sender);
                Gamemode gamemode = this.getToggledGamemode(target);
                if(gamemode == null){
                    throw new CommandException("Unknown gamemode");
                }
                target.setGameMode(gamemode);
            }else{
                throw new CommandException("You are not a player");
            }
        }else if(args.length == 1){
            // /gm 1
            if(sender instanceof EntityPlayer){
                Gamemode gamemode = this.getGameModeFromCommand(args[0]); //TODO: bulletproof this for nulls
                Player target = NailedAPI.getPlayerRegistry().getPlayer((EntityPlayer) sender);
                target.setGameMode(gamemode);
            }else{
                throw new CommandException("You are not a player");
            }
        }else if(args.length == 2){
            // /gm 1 username
            EntityPlayerMP[] matches = getPlayersList(sender, args[1]);
            Gamemode newmode = this.getGameModeFromCommand(args[0]);
            for(EntityPlayerMP match : matches){
                Player target = NailedAPI.getPlayerRegistry().getPlayer(match);
                if(target == null){
                    continue;
                }
                target.setGameMode(newmode);
            }
        }
    }

    protected Gamemode getGameModeFromCommand(String mode) {
        if("survival".equalsIgnoreCase(mode) || "s".equalsIgnoreCase(mode) || "0".equals(mode)){
            return Gamemode.SURVIVAL;
        }else if("creative".equalsIgnoreCase(mode) || "c".equalsIgnoreCase(mode) || "1".equals(mode)){
            return Gamemode.CREATIVE;
        }else if("adventure".equalsIgnoreCase(mode) || "a".equalsIgnoreCase(mode) || "2".equals(mode)){
            return Gamemode.ADVENTURE;
        }
        return null;
    }

    protected Gamemode getToggledGamemode(Player player) {
        if(player.getGameMode() == Gamemode.CREATIVE){
            return Gamemode.SURVIVAL;
        }else{
            return Gamemode.CREATIVE;
        }
    }

    public List<String> addAutocomplete(ICommandSender sender, String[] args) {
        if(args.length == 1){
            return getOptions(args, "survival", "creative", "adventure");
        }else if(args.length == 2){
            return getUsernameOptions(args);
        }
        return null;
    }
}

File: Server/src/main/scala/jk_5/nailed/server/command/CommandGamemode.scala
package jk_5.nailed.server.command

import net.minecraft.command.ICommandSender
import jk_5.nailed.api.map.Map
import net.minecraft.entity.player.EntityPlayer
import jk_5.nailed.api.player.Player
import jk_5.nailed.api.{Gamemode, NailedAPI}

/**
 * No description given
 *
 * @author jk-5
 */
object CommandGamemode extends ScalaCommand {

  val name = "gamemode"
  val usage = "/gamemode <survival|creative|adventure> [player] - Changes your or another player's gamemode"
  override val aliases = Array("gm")

  override def processCommandWithMap(sender: ICommandSender, map: Map, args: Array[String]){
    if(args.length == 0) sender match {
      case e: EntityPlayer => toggleGamemode(NailedAPI.getPlayerRegistry.getPlayer(e))
      case _ => throw new CommandException("You are not a player")
    }else if(args.length == 1) sender match {
      case e: EntityPlayer =>
        val mode = fromString(args(0))
        if(mode.isEmpty) throw new CommandException("Unknown gamemode")
        NailedAPI.getPlayerRegistry.getPlayer(e).setGameMode(mode.get)
      case _ => throw new CommandException("You are not a player")
    }else if(args.length == 2){
      val players = NailedCommand.getPlayersList(sender, args(1))
      val newmode = fromString(args(0))
      if(newmode.isEmpty) throw new CommandException("Unknown gamemode")
      for(p <- players){
        val target = NailedAPI.getPlayerRegistry.getPlayer(p)
        if(target != null) target.setGameMode(newmode.get)
      }
    }
  }

  override def addAutocomplete(sender: ICommandSender, args: Array[String]) = args.length match {
    case 1 => getOptions(args, "survival", "creative", "adventure")
    case 2 => getUsernameOptions(args)
    case _ => null
  }

  def fromString(s: String) =
    if(s.equalsIgnoreCase("survival") || s.equalsIgnoreCase("s") || s == "0") Some(Gamemode.SURVIVAL)
    else if(s.equalsIgnoreCase("creative") || s.equalsIgnoreCase("c") || s == "1") Some(Gamemode.CREATIVE)
    else if(s.equalsIgnoreCase("adventure") || s.equalsIgnoreCase("a") || s == "2") Some(Gamemode.ADVENTURE)
    else None

  def toggleGamemode(p: Player) = p.setGameMode(if(p.getGameMode == Gamemode.SURVIVAL) Gamemode.CREATIVE else Gamemode.SURVIVAL)
}
